### PR TITLE
changed dcm to quaternion function

### PR DESCRIPTION
**Pull request by ryanelandt**
For certain matrices the current function gives the wrong answer. For example:
`atMat(x) = [sin(x) 0 cos(x); 0 -1 0; cos(x) 0 -sin(x)]`
`m1 = atMat(3.1)`
[ 0.0415807 0.0 -0.999135;
0.0 -1.0 0.0;
-0.999135 0.0 -0.0415807]
`println(det(m1))` # 1.0
`println(abs(sum(Quat(m1) - m1)))` # 3.9965406116275517

### CODE REVIEW

File: src/quaternion_types.jl [hunk 1]
```diff
@@ -43,10 +43,45 @@ end
...
+  January 2007
+
+  https://arxiv.org/pdf/math/0701759.pdf
+  This solves the system of equations in Section 3.1
+  """
+
+  a = 1.0 + t[1] + t[5] + t[9]
+  b = 1.0 + t[1] - t[5] - t[9]
+  c = 1.0 - t[1] + t[5] - t[9]
+  d = 1.0 - t[1] - t[5] + t[9]
```
Comment (c42f) [inline src/quaternion_types.jl]: For this to work well with any `eltype(t)`, you should avoid `Float64` literals like `1.0` and `0.5` as these will force conversion to `Float64` even if the input is `Float32`.  It's probably mostly sufficient to use an integer literal instead

```
a = 1 + t[1] + t[5] + t[9]
```

Also, with `divTerm` below, better to use

```
divTerm = 1 / (2*sqrt(a))
```

for the same reasons.

**Comment by c42f** [comment]
Ah, I read the code before your comment which contains one of the problematic cases. Please add it to the unit tests :-)

File: src/quaternion_types.jl [hunk 2]
```diff
@@ -43,10 +43,45 @@ end
...
+    b = t[2] + t[4]
+    divTerm = 0.5 / sqrt(c)
+    d = t[6] + t[8]
+  else  # iMax == 4
+    a = t[2] - t[4]
+    b = t[7] + t[3]
+    c = t[6] + t[8]
+    divTerm = 0.5 / sqrt(d)
+  end
+  return Q(a * divTerm, b * divTerm, c * divTerm, d * divTerm)
```
Comment (tkoolen) [inline src/quaternion_types.jl]: Are the arguments guaranteed to represent a normalized quaternion already? If so, you could skip renormalization inside the `Quat` constructor by passing in `false` as the fifth argument.

File: src/quaternion_types.jl [hunk 3]
```diff
@@ -43,10 +43,45 @@ end
...
+  This solves the system of equations in Section 3.1
+  """
+
+  a = 1.0 + t[1] + t[5] + t[9]
+  b = 1.0 + t[1] - t[5] - t[9]
+  c = 1.0 - t[1] + t[5] - t[9]
+  d = 1.0 - t[1] - t[5] + t[9]
+  maxABCD = max(a, b, c, d)
+  if a == maxABCD
+    divTerm = 0.5 / sqrt(a)
```
Comment (tkoolen) [inline src/quaternion_types.jl]: Rotations has so far followed the standard Julia code formatting guideline for variable names, i.e. all lower case with underscores if necessary. So I think this should be renamed to `divterm` for consistency. Similarly, `maxABCD` -> `max_abcd`.

**Comment by tkoolen** [comment]
Would also be good to add your example as a test case.

File: src/quaternion_types.jl [hunk 4]
```diff
@@ -43,10 +43,45 @@ end
...
+  FROM THE GENERAL FORMULA
+  FOR FOUR-DIMENSIONAL ROTATIONS
+  Johan Ernest Mebius∗
+  January 2007
+
+  https://arxiv.org/pdf/math/0701759.pdf
+  This solves the system of equations in Section 3.1
+  """
+
+  a = 1.0 + t[1] + t[5] + t[9]
```
Comment (andyferris) [inline src/quaternion_types.jl]: `1.0` will e.g. promote `Float32` rotations to `Float64` ones.

(Julia's compiler combined with LLVM are great at optimizing away conversions and it's quite typical just to use `1` in these cases, or else careful use of `one(T)` for appropriate `T`.)

File: src/quaternion_types.jl [hunk 5]
```diff
@@ -43,10 +43,45 @@ end
...
+  d = 1.0 - t[1] - t[5] + t[9]
+  maxABCD = max(a, b, c, d)
+  if a == maxABCD
+    divTerm = 0.5 / sqrt(a)
+    b = t[6] - t[8]
+    c = t[7] - t[3]
+    d = t[2] - t[4]
+  elseif b == maxABCD
+    a = t[6] - t[8]
+    divTerm = 0.5 / sqrt(b)
```
Comment (andyferris) [inline src/quaternion_types.jl]: Similarly, should probably avoid using `0.5` literal (`Float64`) when you don't know the `eltype`. Could divide by `2` or `convert(T, 0.5)` for appropriate `T`.

File: src/quaternion_types.jl (hunk 2 above)
Comment (ryanelandt) [reply src/quaternion_types.jl]: Good point. I would say yes, as the algorithms I am aware of that extract a quaternions from a corrupted rotation matrix are less cheap.

**Comment by ryanelandt** [comment]
Thank you all for the feedback. I'll incorporate these changes and recommit tomorrow. Some things that come to mind... Is it an issue that the structure of my proposed function will prevent the successful taking of finite difference derivatives? Do we care if Rotations functions work with Sympy?

**Comment by c42f** [comment]
Ideally Rotations should work with the julia automatic differentiation packages, and I think the structure of your function should be ok for that - every branch seems like it would propagate a valid derivative.

If we have to choose between numerical stability and the ability to have symbolic derivatives, I'd certainly choose the numerically stable option.

File: src/quaternion_types.jl [hunk 6]
```diff
@@ -43,12 +43,45 @@ end
 
 # These 3 functions are enough to satisfy the entire StaticArrays interface:
 function (::Type{Q})(t::NTuple{9}) where Q<:Quat
-    q = Q(sqrt(abs(1  + t[1] + t[5] + t[9])) / 2,
-               copysign(sqrt(abs(1 + t[1] - t[5] - t[9]))/2, t[6] - t[8]),
-               copysign(sqrt(abs(1 - t[1] + t[5] - t[9]))/2, t[7] - t[3]),
-               copysign(sqrt(abs(1 - t[1] - t[5] + t[9]))/2, t[2] - t[4]))
+  """
```
Comment (andyferris) [inline src/quaternion_types.jl]: Thanks for the documentation here.

However - docstrings go before the function (or else use comments (`#`) or a comment block (`#= ... =#`))

**Comment by andyferris** [comment]
Cool, thanks.

> Normalization should happen there in order to support Dual numbers.

If the goal is to make sure that constructors compose correctly w.r.t. derivatives, should we just make a test or two using dual numbers or something? Did you play at the REPL to make sure you get the derivatives you expect?

**Comment by ryanelandt** [comment]
I did a little, and I've worked with quaternion and duals before. Although this program (as well as its predecessor) only work for orthonormal matrices, I think that explicitly normalizing after is best. This (at a minimum) ensures that the partials of the returned quaternion are consistent with a unit quaternion (even if the partials of the matrix are not consistent with an orthonormal matrix). As currently written, taking a quaternion of type dual, creating a rotation matrix from this, and then extracting the quaternion from this matrix produces the same quaternion values and partials that you started with. I'm not sure how meaningful the dual results will be for matrices with inconsistent partials, but at the very least normalizing in the constructor guarantees consistent partials.

`using Rotations
using ForwardDiff: Dual
seed = normalize(randn(4))
q1 = Dual(seed[1], 1.0, 0.0, 0.0, 0.0)
q2 = Dual(seed[2], 0.0, 1.0, 0.0, 0.0)
q3 = Dual(seed[3], 0.0, 0.0, 1.0, 0.0)
q4 = Dual(seed[4], 0.0, 0.0, 0.0, 1.0)
quat_unit = Quat(q1, q2, q3, q4, true)
dcm_unit = RotMatrix(quat_unit)
quat_unit_back = Quat(dcm_unit)
the_error = sum(abs.(quat_unit - quat_unit_back))
println(norm([the_error.partials..., the_error.value]))`

**Comment by tkoolen** [comment]
Sorry about the delay; looks good to me.

**Comment by tkoolen** [comment]
I checked the benchmark times BTW; they're almost exactly the same.

File: src/quaternion_types.jl [hunk 7]
```diff
@@ -43,12 +43,45 @@ end
...
+  elseif c == max_abcd
+    a = t[7] - t[3]
+    b = t[2] + t[4]
+    d = t[6] + t[8]
+  else
+    a = t[2] - t[4]
+    b = t[7] + t[3]
+    c = t[6] + t[8]
+  end
+  return Q(a, b, c, d)
```
Comment (c42f) [inline src/quaternion_types.jl]: A nitpicking problem: this indentation style is inconsistent with the rest of the package (four spaces per indentation level)

File: test/rotation_tests.jl [hunk 8]
```diff
@@ -178,6 +178,24 @@ all_types = (RotMatrix{3}, Quat, SPQuat, AngleAxis, RodriguesVec,
         end
     end
 
+    #########################################################################
+    # Test robustness of DCM to Quat function
+    #########################################################################
+    function nearest_orthonormal(not_orthonormal)
+      # Didn't see a function like this already in Rotations
```
Comment (c42f) [inline test/rotation_tests.jl]: Probably can delete this comment - it's more of a PR meta-comment which doesn't add much to future maintainability of the package ;-)

**Comment by tkoolen** [comment]
I'm going to go ahead and merge this since everything seems to be addressed.

**Pull request merged**